pkg/sync/service.go

```diff
@@ -388,6 +388,73 @@ func (s *Service) doDownloadFromURL(request *http.Request) (err error) {
 	return nil
 }
 
+func (s *Service) CloneFromBackingImage(writer http.ResponseWriter, request *http.Request) {
+	err := s.doCloneFromBackingImage(request)
+	if err != nil {
+		http.Error(writer, err.Error(), http.StatusInternalServerError)
+		return
+	}
+}
+
+func (s *Service) doCloneFromBackingImage(request *http.Request) (err error) {
+	defer func() {
+		if err != nil {
+			s.log.WithError(err).Errorf("Sync Service: failed to do clone from backing image")
+		}
+	}()
+
+	queryParams := request.URL.Query()
+	sourceBackingImage := queryParams.Get(types.DataSourceTypeCloneParameterBackingImage)
+	if sourceBackingImage == "" {
+		return fmt.Errorf("%v is not specified", types.DataSourceTypeCloneParameterBackingImage)
+	}
+
+	sourceBackingImageUUID := queryParams.Get(types.DataSourceTypeCloneParameterBackingImageUUID)
+	if sourceBackingImageUUID == "" {
+		return fmt.Errorf("%v is not specified", types.DataSourceTypeCloneParameterBackingImageUUID)
+	}
+
+	encryption := types.EncryptionType(queryParams.Get(types.DataSourceTypeCloneParameterEncryption))
+	if encryption != types.EncryptionTypeEncrypt && encryption != types.EncryptionTypeDecrypt && encryption != types.EncryptionTypeIgnore {
+		return fmt.Errorf("%v operation %v is not specified", types.DataSourceTypeCloneParameterEncryption, encryption)
+	}
+	filePath := queryParams.Get("file-path")
+	if filePath == "" {
+		return fmt.Errorf("no filePath restoring file")
+	}
+	uuid := queryParams.Get("uuid")
+	if uuid == "" {
+		return fmt.Errorf("no uuid for restoring file")
+	}
+	diskUUID := queryParams.Get("disk-uuid")
+	expectedChecksum := queryParams.Get("expected-checksum")
+
+	credential := map[string]string{}
+	if err := json.NewDecoder(request.Body).Decode(&credential); err != nil {
+		if err != io.EOF {
+			return fmt.Errorf("get credential failed from request")
+		}
+	}
+
+	sf, err := s.checkAndInitSyncFile(filePath, uuid, diskUUID, expectedChecksum, 0)
+	if err != nil {
+		return err
+	}
+	go func() {
+		if err := sf.WaitForStateNonPending(); err != nil {
+			s.log.Errorf("Sync Service: failed to wait for sync file %v becoming non-pending state before starting the actual cloning: %v", filePath, err)
+			return
+		}
```

Comment:
What cases can result in the pending state?

Reply:
The init state of the backing file is `pending`
we check if we can reuse the file on the host

you can check the state transition by @shuo-wu 
https://github.com/longhorn/backing-image-manager/blob/master/pkg/sync/sync_file.go#L26
```
// SyncFile state machine:
//
//     +---------+  Try to reuse file (best effort),              +-----------+
//  +--+ pending +----------------------------------------------->+ starting  +-----+
//  |  +----+----+  do cleanup & preparation if reuse failed.     +-----+-----+     |
//  |       |                                                           |           |
//  |       |                                                           |           |
//  |       |                           Sender doesn't start sending    |           |
//  |       |                         +---------------------------------+           |
//  |       |                         | data after an interval.                     |
//  |       |                         |                                             |
//  |       |                         |                               Start syncing |
//  |       |                         |                               the file.     |
//  |       |                         v                                             |
//  |       |                  +------+------+                                      |
//  |       +----------------->+   failed    +<--------------------------+          |
//  |        Fail at cleanup   +------+------+                           |          |
//  |        or preparation.          ^                                  |          |
//  |                                 |           Fail at file handling, |          |
//  |  Reuse success.                 |                  data transport, |          |
//  |                                 |         or the final validation. |          |
//  |                                 |                                  |          |
//  |   +-------------+               |                                  |          |
//  |   +   unknown   + --------------+                                  |          |
//  |   +------+------+  Checksum is changed after modification.         |          |
//  |          ^                                                         |          |
//  |          | File is suddenly modified.                              |          |
//  |          | Need to re-calculate the checksum.                      |          |
//  |          v                                                         |          |
//  |   +-------------+                                           +------+------+   |
//  +-->+    ready    + <-----------------------------------------+ in-progress +<--+
//      +------+------+                                           +------+------+
//
```